### PR TITLE
pass all string decoder tests

### PR DESCRIPTION
**Pull request by paperclover**
originally, i wanted to use bindgen for the string decoder class.

instead, i just fixed the bugs. this class is a mess. ideally after this pr the code wont have to be edited since it is now 100% passing, and its a legacy api node hasn't touched in forever.

### CODE REVIEW

**Comment by robobun** [comment]
@paperdave, your commit c27f28876d616189a8643d41f3af1de40bcc1ee8 is building: [`#7865`](https://buildkite.com/bun/bun/builds/7865)

File: src/bun.js/bindings/ErrorCode.cpp [hunk 1]
```diff
@@ -302,16 +302,9 @@ WTF::String ERR_OUT_OF_RANGE(JSC::ThrowScope& scope, JSC::JSGlobalObject* global
 
 namespace ERR {
 
-JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::String& val_arg_name, const WTF::String& val_expected_type, JSC::JSValue val_actual_value)
+JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::ASCIILiteral& arg_name, const WTF::ASCIILiteral& expected_type, JSC::JSValue val_actual_value)
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/ErrorCode.cpp]: asciiliteral doesn't need const &

```suggestion
JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, WTF::ASCIILiteral arg_name, WTF::ASCIILiteral expected_type, JSC::JSValue val_actual_value)
```

File: src/bun.js/bindings/ErrorCode.cpp [hunk 2]
```diff
@@ -322,15 +315,11 @@ JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalO
     throwScope.throwException(globalObject, createError(globalObject, ErrorCode::ERR_INVALID_ARG_TYPE, message));
     return {};
 }
-JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, JSC::JSValue val_arg_name, const WTF::String& val_expected_type, JSC::JSValue val_actual_value)
+JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, JSC::JSValue val_arg_name, const WTF::ASCIILiteral& expected_type, JSC::JSValue val_actual_value)
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/ErrorCode.cpp]:
```suggestion
JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, JSC::JSValue val_arg_name, WTF::ASCIILiteral expected_type, JSC::JSValue val_actual_value)
```

File: src/bun.js/bindings/ErrorCode.cpp [hunk 3]
```diff
@@ -408,15 +397,9 @@ JSC::EncodedJSValue OUT_OF_RANGE(JSC::ThrowScope& throwScope, JSC::JSGlobalObjec
     return {};
 }
 
-JSC::EncodedJSValue INVALID_ARG_VALUE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::String& name, JSC::JSValue value, const WTF::String& reason)
+JSC::EncodedJSValue INVALID_ARG_VALUE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::ASCIILiteral& name, JSC::JSValue value, const WTF::String& reason)
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/ErrorCode.cpp]:
```suggestion
JSC::EncodedJSValue INVALID_ARG_VALUE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, WTF::ASCIILiteral name, JSC::JSValue value, const WTF::String& reason)
```

File: src/bun.js/bindings/ErrorCode.h [hunk 4]
```diff
@@ -75,14 +75,14 @@ enum Bound {
...
 JSC::EncodedJSValue OUT_OF_RANGE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, JSC::JSValue arg_name, size_t lower, size_t upper, JSC::JSValue actual);
 JSC::EncodedJSValue OUT_OF_RANGE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, JSC::JSValue arg_name_val, size_t bound_num, Bound bound, JSC::JSValue actual);
 JSC::EncodedJSValue OUT_OF_RANGE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, JSC::JSValue arg_name_val, const WTF::String& msg, JSC::JSValue actual);
 JSC::EncodedJSValue OUT_OF_RANGE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::String& arg_name_val, const WTF::String& msg, JSC::JSValue actual);
-JSC::EncodedJSValue INVALID_ARG_VALUE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::String& name, JSC::JSValue value, const WTF::String& reason = "is invalid"_s);
+JSC::EncodedJSValue INVALID_ARG_VALUE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::ASCIILiteral& name, JSC::JSValue value, const WTF::String& reason = "is invalid"_s);
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/ErrorCode.h]:
```suggestion
JSC::EncodedJSValue INVALID_ARG_VALUE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, WTF::ASCIILiteral name, JSC::JSValue value, const WTF::String& reason = "is invalid"_s);
```

File: src/bun.js/bindings/ErrorCode.h [hunk 5]
```diff
@@ -75,14 +75,14 @@ enum Bound {
 
 namespace ERR {
 
-JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::String& val_arg_name, const WTF::String& val_expected_type, JSC::JSValue val_actual_value);
-JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, JSC::JSValue val_arg_name, const WTF::String& val_expected_type, JSC::JSValue val_actual_value);
+JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::ASCIILiteral& arg_name, const WTF::ASCIILiteral& expected_type, JSC::JSValue val_actual_value);
+JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, JSC::JSValue arg_name, const WTF::ASCIILiteral& expected_type, JSC::JSValue val_actual_value);
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/ErrorCode.h]:
```suggestion
JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, JSC::JSValue arg_name, WTF::ASCIILiteral expected_type, JSC::JSValue val_actual_value);
```

File: src/bun.js/bindings/ErrorCode.h [hunk 6]
```diff
@@ -75,14 +75,14 @@ enum Bound {
 
 namespace ERR {
 
-JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::String& val_arg_name, const WTF::String& val_expected_type, JSC::JSValue val_actual_value);
-JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, JSC::JSValue val_arg_name, const WTF::String& val_expected_type, JSC::JSValue val_actual_value);
+JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::ASCIILiteral& arg_name, const WTF::ASCIILiteral& expected_type, JSC::JSValue val_actual_value);
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/ErrorCode.h:78]:
```suggestion
JSC::EncodedJSValue INVALID_ARG_TYPE(JSC::ThrowScope& throwScope, JSC::JSGlobalObject* globalObject, const WTF::ASCIILiteral& arg_name, WTF::ASCIILiteral expected_type, JSC::JSValue val_actual_value);
```

**Review by Jarred-Sumner — APPROVED** [review]
LGTM

Extremely minor nitpicks

File: test/js/node/test/parallel/test-string-decoder-fuzz.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  strings.push(sd.end());
+
+  assert.strictEqual(strings.join(''), Buffer.concat(bufs).toString(enc),
+                     `Mismatch:\n${util.inspect(strings)}\n` +
+                     util.inspect(bufs.map((buf) => buf.toString('hex'))) +
+                     `\nfor encoding ${enc}`);
+}
+
+const start = Date.now();
+while (Date.now() - start < 10000)
```
Comment (Jarred-Sumner) [inline test/js/node/test/parallel/test-string-decoder-fuzz.js]: 10s is too long

```suggestion
// Run this for 1 second
while (Date.now() - start < 1000)
```

**Pull request merged**